File: MasterCollateralLibrary/Scripts/app/clientAccounts/controllers/clientAccountsController.js


(function () {
    'use strict';

    angular
        .module('app.clientAccounts')
        .controller('clientAccountsController', ['$scope', '$q', '$location', 'clientAccountsService',
            'quoteService', 'distributorService', 'gpoService', 'alertService', clientAccountsController]);

    function clientAccountsController($scope, $q, $location, clientAccountsService,
        quoteService, distributorService, gpoService, alertService) {

        $scope.clientAccount = {
            id: 0,
            acct_name: '',
            acct_city: '',
            acct_state: '',
            acct_cntry: '',
            acct_3code: '',
            acct_whrhs: '',
            acct_frght: '',
            acct_misc: '',
            acct_logstcs: '',
            acct_admin: '',
            acct_contract_start: null,
            acct_contract_end: null,
            acct_assoc: null,
            acct_gpo_id: null,
            acct_auth_dist_id: null,
            acct_cat_id: null,
            acct_marketing_percent_fee: null,
            acct_gpo_percent_fee: null,
            acct_spif_percent_fee: null,
            acct_dist_street1: '',
            acct_dist_street2: '',
            acct_dist_city: '',
            acct_dist_state: '',
            acct_dist_zip: '',
            acct_dist_country: '',
            acct_rep_id: null,
            acct_address: '',
            acct_zip: '',
            acct_primary_contact_name: '',
            acct_primary_contact_dept: '',
            acct_primary_contact_phone: '',
            acct_primary_contact_email: '',
            acct_secondary_contact_name: '',
            acct_secondary_contact_dept: '',
            acct_secondary_contact_phone: '',
            acct_secondary_contact_email: ''
        };

        $scope.errors = [];

        var temp = window.location.pathname.split('/');
        var id = 0;
        if (temp.length === 4) {
            id = temp[3];
        }

        if (id > 0) {
            clientAccountsService.getClientAccountEdit(id)
                .then(function(response) {
                    if (response !== "" && response !== undefined) {
                        if (response.acct_contract_start !== null) {
                            response.acct_contract_start = new Date(parseInt(response.acct_contract_start.substr(6)));
                        }
                        if (response.acct_contract_end !== null) {
                            response.acct_contract_end = new Date(parseInt(response.acct_contract_end.substr(6)));
                        }
                        $scope.clientAccount = response;
                        getDropDownListData();
                    }
                });
        } else {
            getDropDownListData();
        }

        
        function getDropDownListData() {
            $scope.selectedCompany = null;
            getBrands();
            getCategories();
            getDistributors();
            getGPOs();
            getReps();
        }

        $scope.brands = [];
        $scope.selectedCompany = null;        
        function getBrands() {
            quoteService.getBrands()
                .then(function (res) {
                    $scope.brands = res;
                    //console.log(res);
                    if ($scope.clientAccount === undefined) return;
                    var brandFound = false;

                    for (var i = 0; i < res.length; i++) {
                        if (res[i].id === $scope.clientAccount.acct_assoc) {
                            $scope.selectedCompany = res[i];
                            brandFound = true;
                        }
                        if (brandFound === true) break;
                    }
                });
        }

        $scope.categories = [];
        $scope.selectedCategory = null;
        function getCategories() {
            clientAccountsService.getAccountCategories()
                .then(function (response) {
                    $scope.categories = response;
                    if (!$scope.clientAccount
                        || $scope.clientAccount.id < 1
                        || $scope.clientAccount.acct_cat_id === null) {
                        return;
                    }
                    for (var i = 0; i < response.length; i++) {
                        if (response[i].id === $scope.clientAccount.acct_cat_id) {
                            $scope.selectedCategory = response[i];
                            break;
                        }                        
                    }
                });
        }

        $scope.distributors = [];
        $scope.selectedDistributor = null;
        function getDistributors() {
            distributorService.getDistributors()
                .then(function (response) {
                    $scope.distributors = response;
                    if ($scope.clientAccount.id === 0
                        || $scope.clientAccount.acct_auth_dist_id === null) {
                        return;
                    }
                    for (var i = 0; i < response.length; i++) {
                        if ($scope.clientAccount.acct_auth_dist_id === response[i].id) {
                            $scope.selectedDistributor = response[i];                            
                            break;
                        }
                    }
                });
        }

        $scope.gpos = [];
        $scope.selectedGPO = null;
        function getGPOs() {
            gpoService.getGPOs()
                .then(function (response) {
                    $scope.gpos = response;
                    if ($scope.clientAccount.id === 0
                        || $scope.clientAccount.acct_gpo_id === null) {
                        return;
                    }
                    for (var i = 0; i < response.length; i++) {
                        if ($scope.clientAccount.acct_gpo_id === response[i].id) {
                            $scope.selectedGPO = response[i];
                            break;
                        }
                    }
                });
        }

        $scope.reps = [];
        $scope.selectedRep = null;
        function getReps() {
            quoteService.getSalesReps()
                .then(function (response) {
                    $scope.reps = response;
                    if ($scope.clientAccount.id < 1) return;
                    for (var i = 0; i < response.length; i++) {
                        if (response[i].id === $scope.clientAccount.acct_rep_id) {
                            $scope.selectedRep = response[i];
                            break;
                        }
                    }
                });
        }

        $scope.save = function () {
            if ($scope.selectedCompany !== null) {
                $scope.clientAccount.acct_assoc = $scope.selectedCompany.id;
            }
            if ($scope.selectedCategory !== null) {
                $scope.clientAccount.acct_cat_id = $scope.selectedCategory.id;
            }
            if ($scope.selectedDistributor !== null) {
                $scope.clientAccount.acct_auth_dist_id = $scope.selectedDistributor.id;
            }
            if ($scope.selectedGPO !== null) {
                $scope.clientAccount.acct_gpo_id = $scope.selectedGPO.id;
            }
            if ($scope.selectedRep !== null) {
                $scope.clientAccount.acct_rep_id = $scope.selectedRep.id;
            }
            if ($scope.clientAccount.acct_address !== null) {
                $scope.clientAccount.acct_address = $scope.clientAccount.acct_address;
            }
            if ($scope.clientAccount.acct_zip !== null) {
                $scope.clientAccount.acct_zip = $scope.clientAccount.acct_zip;
            }
            // validation
            $scope.errors = [];
            if ($scope.clientAccount.acct_3code.length < 1) {
                $scope.errors.push('Account Code is required.');
                return;
            }

            if ($scope.errors.length > 0) {
                return;
            }

            // validate dup acct code
            clientAccountsService.isAccountCodeExists($scope.clientAccount.acct_3code, $scope.clientAccount.id)
                .then(function (response) {
                    console.log('val', response);
                    if (response === false) {
                        // update
                        if ($scope.clientAccount.id > 0) {
                            clientAccountsService.updateClientAccount($scope.clientAccount)
                                .then(function (res) {
                                    if (res === true) {
                                        //Show result notification
                                        window.location.href = "/ClientAccounts/";
                                    }
                                });
                        } else { // create
                            clientAccountsService.createClientAccount($scope.clientAccount)
                                .then(function (res) {
                                    if (res.id > 0) {
                                        window.location.href = "/ClientAccounts/Edit/" + res.id;
                                    }
                                });
                        }                        
                    } else {
                        $scope.errors.push('Account Code already exists.');
                    }
                });
        }

        $scope.cancel = function () {

            window.location = window.location.href = "/ClientAccounts/";
        }        
    }
    

})();

File: MasterCollateralLibrary/Controllers/ClientAccountsController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using AutoMapper;
using MCL.DTOs;
using MCL.EF.DAL;
using MCL.EF.DAL.Interfaces;
using MCL.EF.Model;
using MCL.Services;
using NLog;

namespace MasterCollateralLibrary.Controllers
{
    [Authorize]
    public class ClientAccountsController : Controller
    {
        protected readonly IUnitOfWork UnitOfWork;
        private static Logger _logger = LogManager.GetCurrentClassLogger();

        public ClientAccountsController()
        {
            UnitOfWork = new UnitOfWork(new GRI_DBEntities());
        }

        // GET: Brands
        public ActionResult Index()
        {
            return View();
        }

        public ActionResult Edit(int id)
        {
            return View();
        }

        public ActionResult Create()
        {
            return View("Edit");
        }

        [HttpGet]
        public JsonResult GetRepEdit(int id)
        {
            var data = UnitOfWork.TblAccount.Get(id);
            var dto = Mapper.Map<AccountDTO>(data);
            return Json(dto, JsonRequestBehavior.AllowGet);
        }

        [HttpPost]
        public JsonResult Find(AccountFindDTO param)
        {
            var data = UnitOfWork.TblAccount.Find(param.PageSize, param.PageNumber,
                param.SortBy, param.SortDir, param.acct_name, param.acct_city, param.acct_state);
            var dto = Mapper.Map<PageListDTO<ICollection<AccountFindDTO>>>(data);
            return Json(dto, JsonRequestBehavior.AllowGet);
        }

        [HttpPut]
        public JsonResult Update(AccountDTO dto)
        {
            bool succeeded = false;

            try
            {
                var model = UnitOfWork.TblAccount.Get(dto.id);

                if (model != null)
                {
                    model.acct_name = dto.acct_name;
                    model.acct_city = dto.acct_city;
                    model.acct_state = dto.acct_state;
                    model.acct_cntry = dto.acct_cntry;
                    model.acct_3code = dto.acct_3code;
                    model.acct_whrhs = dto.acct_whrhs;
                    model.acct_frght = dto.acct_frght;
                    model.acct_misc = dto.acct_misc;
                    model.acct_logstcs = dto.acct_logstcs;
                    model.acct_admin = dto.acct_admin;
                    model.acct_contract_start = dto.acct_contract_start;
                    model.acct_contract_end = dto.acct_contract_end;
                    model.acct_assoc = dto.acct_assoc;
                    model.acct_gpo_id = dto.acct_gpo_id;
                    model.acct_auth_dist_id = dto.acct_auth_dist_id;
                    model.acct_cat_id = dto.acct_cat_id;
                    model.acct_marketing_percent_fee = dto.acct_marketing_percent_fee;
                    model.acct_gpo_percent_fee = dto.acct_gpo_percent_fee;
                    model.acct_spif_percent_fee = dto.acct_spif_percent_fee;
                    model.acct_dist_street1 = dto.acct_dist_street1;
                    model.acct_dist_street2 = dto.acct_dist_street2;
                    model.acct_dist_city = dto.acct_dist_city;
                    model.acct_dist_state = dto.acct_dist_state;
                    model.acct_dist_zip = dto.acct_dist_zip;
                    model.acct_dist_country = dto.acct_dist_country;
                    model.acct_rep_id = dto.acct_rep_id;
                    model.acct_address = dto.acct_address;
                    model.acct_zip = dto.acct_zip;
                    model.acct_primary_contact_name = dto.acct_primary_contact_name;
                    model.acct_primary_contact_dept = dto.acct_primary_contact_dept;
                    model.acct_primary_contact_phone = dto.acct_primary_contact_phone;
                    model.acct_primary_contact_email = dto.acct_primary_contact_email;
                    model.acct_secondary_contact_name = dto.acct_secondary_contact_name;
                    model.acct_secondary_contact_dept = dto.acct_secondary_contact_dept;
                    model.acct_secondary_contact_phone = dto.acct_secondary_contact_phone;
                    model.acct_secondary_contact_email = dto.acct_secondary_contact_email;

                    UnitOfWork.Save();
                    succeeded = true;
                }
            }
            catch (Exception ex)
            {
                succeeded = false;
                _logger.Error(ex);
            }
            return Json(succeeded, JsonRequestBehavior.AllowGet);
        }

        [HttpPost]
        public JsonResult SaveNew(AccountDTO dto)
        {
            try
            {
                var model = new tbl_acct();

                model.acct_name = dto.acct_name;
                model.acct_city = dto.acct_city;
                model.acct_state = dto.acct_state;
                model.acct_cntry = dto.acct_cntry;
                model.acct_3code = dto.acct_3code;
                model.acct_whrhs = dto.acct_whrhs;
                model.acct_frght = dto.acct_frght;
                model.acct_misc = dto.acct_misc;
                model.acct_logstcs = dto.acct_logstcs;
                model.acct_admin = dto.acct_admin;
                model.acct_contract_start = dto.acct_contract_start;
                model.acct_contract_end = dto.acct_contract_end;
                model.acct_assoc = dto.acct_assoc;
                model.acct_gpo_id = dto.acct_gpo_id;
                model.acct_auth_dist_id = dto.acct_auth_dist_id;
                model.acct_cat_id = dto.acct_cat_id;
                model.acct_marketing_percent_fee = dto.acct_marketing_percent_fee;
                model.acct_gpo_percent_fee = dto.acct_gpo_percent_fee;
                model.acct_spif_percent_fee = dto.acct_spif_percent_fee;
                model.acct_dist_street1 = dto.acct_dist_street1;
                model.acct_dist_street2 = dto.acct_dist_street2;
                model.acct_dist_city = dto.acct_dist_city;
                model.acct_dist_state = dto.acct_dist_state;
                model.acct_dist_zip = dto.acct_dist_zip;
                model.acct_dist_country = dto.acct_dist_country;

                UnitOfWork.TblAccount.Add(model);

                UnitOfWork.Save();
                dto.id = model.id;
            }
            catch (Exception ex)
            {
                dto.id = 0;
                _logger.Error(ex);
            }
            return Json(dto, JsonRequestBehavior.AllowGet);
        }

        [HttpDelete]
        public JsonResult Delete(int id)
        {
            bool succeeded = false;

            try
            {
                var model = UnitOfWork.TblAccount.Get(id);

                if (model != null)
                {
                    UnitOfWork.TblAccount.Remove(model);
                    UnitOfWork.Save();
                    succeeded = true;
                }
            }
            catch (Exception ex)
            {
                succeeded = false;
                _logger.Error(ex);
            }
            return Json(succeeded, JsonRequestBehavior.AllowGet);
        }

        [HttpGet]
        public JsonResult GetBrands()
        {
            var data = UnitOfWork.TblBrand.GetAll()
                .OrderBy(t => t.brand_name)
                .ToList();

            var dto = Mapper.Map<IEnumerable<BrandDTO>>(data);
            return Json(dto, JsonRequestBehavior.AllowGet);
        }

        [HttpGet]
        public JsonResult GetAccountCategories()
        {
            var data = UnitOfWork.TblAccountCategory.GetAll()
                .Where(t => t.cat_deleted == false)
                .OrderBy(t => t.cat_name)
                .ToList();
            var dto = Mapper.Map<IEnumerable<AccountCategoryDTO>>(data);
            return Json(dto, JsonRequestBehavior.AllowGet);
        }

        [HttpPost]
        public JsonResult IsAccountCodeExists(string accountCode, int excludeId)
        {
            var result = false;

            try
            {
                result = UnitOfWork.TblAccount.IsAccountCodeExists(accountCode, excludeId);                    
            }
            catch (Exception ex)
            {
                _logger.Error(ex);
            }
            return Json(result, JsonRequestBehavior.AllowGet);
        }
    }
}